src/cli.rs

```diff
@@ -0,0 +1,764 @@
+use anyhow::anyhow;
+use clap::{
+    builder::RangedU64ValueParser, value_parser, AppSettings, Arg, ArgAction, ArgGroup, ArgMatches,
+    Command, ErrorKind, Parser, ValueEnum,
+};
+#[cfg(feature = "completions")]
+use clap_complete::Shell;
+use normpath::PathExt;
+
+use crate::error::print_error;
+use crate::exec::CommandSet;
+use crate::filesystem;
+#[cfg(unix)]
+use crate::filter::OwnerFilter;
+use crate::filter::SizeFilter;
+
+// Type for options that don't have any values, but are used to negate
+// earlier options
+struct Negations;
+
+impl clap::FromArgMatches for Negations {
+    fn from_arg_matches(_: &ArgMatches) -> clap::Result<Self> {
+        Ok(Negations)
+    }
+
+    fn update_from_arg_matches(&mut self, _: &ArgMatches) -> clap::Result<()> {
+        Ok(())
+    }
+}
+
+impl clap::Args for Negations {
+    fn augment_args(cmd: Command<'_>) -> Command<'_> {
+        Self::augment_args_for_update(cmd)
+    }
+
+    fn augment_args_for_update(cmd: Command<'_>) -> Command<'_> {
+        cmd.arg(
+            Arg::new("no-hidden")
+                .long("no-hidden")
+                .overrides_with("hidden")
+                .hide(true)
+                .long_help("Overrides --hidden."),
+        )
+        .arg(
+            Arg::new("ignore")
+                .long("ignore")
+                .overrides_with("no-ignore")
+                .hide(true)
+                .long_help("Overrides --no-ignore."),
+        )
+        .arg(
+            Arg::new("ignore-vcs")
+                .long("ignore-vcs")
+                .overrides_with("no-ignore-vcs")
+                .hide(true)
+                .long_help("Overrides --no-ignore-vcs."),
+        )
+        .arg(
+            Arg::new("relative-path")
+                .long("relative-path")
+                .overrides_with("absolute-path")
+                .hide(true)
+                .long_help("Overrides --absolute-path."),
+        )
+        .arg(
+            Arg::new("no-follow")
+                .long("no-follow")
+                .overrides_with("follow")
+                .hide(true)
+                .long_help("Overrides --follow."),
+        )
+    }
+}
+
+#[derive(Parser)]
+#[clap(
+    version,
+    setting(AppSettings::DeriveDisplayOrder),
+    dont_collapse_args_in_usage = true,
+    after_help = "Note: `fd -h` prints a short and concise overview while `fd --help` gives all \
+    details.",
+    group(ArgGroup::new("execs").args(&["exec", "exec-batch", "list-details"]).conflicts_with_all(&[
+            "max-results", "has-results", "count"])),
+)]
+pub struct Opts {
+    /// Search hidden files and directories
+    ///
+    /// Include hidden directories and files in the search results (default:
+    /// hidden files and directories are skipped). Files and directories are considered
+    /// to be hidden if their name starts with a `.` sign (dot).
+    /// The flag can be overridden with --no-hidden.
+    #[clap(long, short = 'H', action, overrides_with = "hidden")]
```

Comment:
If you opt-in to clap v4 behavior with a bald `action` attribute, args override themselves and you no longer need `overrides_with = "<self>"`

Reply:
It looks like this has now changed to the opposite? from the v4 changelog:

> ArgAction::Set, ArgAction::SetTrue, and Arg::Action::SetFalse now conflict by default